src/store.tsx

```diff
@@ -0,0 +1,117 @@
+import { useReducer, useEffect, createContext, useContext, useCallback, useMemo, useState } from "react";
+
+export interface Movie {
+  id: string;
+  name: string;
+  genres: string[];
+  year: number;
+  poster: string;
+}
+
+function useMoviesSource(): {
+  movies: Movie[];
+  filterBy: string;
+  setFilterBy: (filterBy: string) => void;
+  sortBy: string;
+  setSortBy: (sortBy: string) => void;
+  selectedMovie: Movie | undefined;
+  setSelectedMovieId: (movieId: string) => void;
+} {
+  type MoviesState = {
+    movies: Movie[];
+    filterBy: string;
+    sortBy: string;
+  };
+
+  type MoviesAction =
+    | { type: "setMovies"; payload: Movie[] }
+    | { type: "setFilterBy"; payload: string }
+    | { type: "setSortBy"; payload: string };
+
+  const [{ movies, filterBy, sortBy }, dispatch] = useReducer(
+    (state: MoviesState, action: MoviesAction) => {
+      switch (action.type) {
+        case "setMovies":
+          return { ...state, movies: action.payload };
+        case "setFilterBy":
+          return { ...state, filterBy: action.payload };
+        case "setSortBy":
+          return { ...state, sortBy: action.payload };
+      }
+    },
+    {
+      movies: [],
+      filterBy: "",
+      sortBy: "",
+    }
+  );
+
+  useEffect(() => {
+    fetch("/data/movies.json")
+      .then((response) => response.json())
+      .then((data) =>
+        dispatch({
+          type: "setMovies",
+          payload: data,
+        })
+      );
+  }, []);
+
+  const [selectedMovie, setSelectedMovie] = useState<Movie | undefined>(undefined);
+
+  const setFilterBy = useCallback((filterBy: string) => {
+    dispatch({
+      type: "setFilterBy",
+      payload: filterBy,
+    });
+  }, []);
+
+  const setSortBy = useCallback((sortBy: string) => {
+    dispatch({
+      type: "setSortBy",
+      payload: sortBy,
+    });
+  }, []);
+
+  const setSelectedMovieId = useCallback(
+    (movieId: string) => {
+      console.log(movieId);
+      const movie = movies.find((movie) => movie.id === movieId);
+      setSelectedMovie(movie);
+    },
+    [setSelectedMovie, movies]
+  );
```

Comment:
Please read this beautiful article here: https://kentcdodds.com/blog/usememo-and-usecallback
UseCallbacks are appropriate only when you pass a function as a prop to another component. So, you do not need it here :(

Reply:
Not Resolved 👀